mobile_verifier/src/boosting_oracles/data_sets.rs

```diff
@@ -0,0 +1,482 @@
+use std::{
+    collections::HashMap,
+    path::{Path, PathBuf},
+    pin::pin,
+    sync::Arc,
+};
+
+use chrono::{DateTime, Duration, Utc};
+use file_store::{file_sink::FileSinkClient, traits::TimestampEncode, FileStore};
+use futures_util::{Stream, StreamExt, TryFutureExt, TryStreamExt};
+use helium_proto::services::poc_mobile as proto;
+use rust_decimal::prelude::ToPrimitive;
+use rust_decimal_macros::dec;
+use sqlx::{FromRow, PgPool, QueryBuilder};
+use task_manager::ManagedTask;
+use tokio::{fs::File, io::AsyncWriteExt, sync::Mutex};
+
+use crate::{boosting_oracles::assignment::HexAssignments, coverage::SignalLevel};
+
+use super::{HexAssignment, HexBoostData};
+
+pub trait DataSet: HexAssignment + Send + Sync + 'static {
+    const TYPE: DataSetType;
+
+    fn timestamp(&self) -> Option<DateTime<Utc>>;
+
+    fn update(&mut self, path: &Path, time_to_use: DateTime<Utc>) -> anyhow::Result<()>;
+
+    fn is_ready(&self) -> bool;
+}
+
+pub struct DataSetDownloaderDaemon<T, A, B, C> {
+    pool: PgPool,
+    data_set: Arc<Mutex<T>>,
+    data_sets: HexBoostData<A, B, C>,
+    store: FileStore,
+    oracle_boosting_sink: FileSinkClient,
+    data_set_directory: PathBuf,
+    latest_file_date: Option<DateTime<Utc>>,
+}
+
+#[derive(FromRow)]
+pub struct NewDataSet {
+    filename: String,
+    time_to_use: DateTime<Utc>,
+    status: DataSetStatus,
+}
+
+#[derive(Copy, Clone, sqlx::Type)]
+#[sqlx(type_name = "data_set_status")]
+#[sqlx(rename_all = "lowercase")]
+pub enum DataSetStatus {
+    Pending,
+    Downloaded,
+    Processed,
+}
+
+impl DataSetStatus {
+    pub fn is_downloaded(&self) -> bool {
+        matches!(self, Self::Downloaded)
+    }
+}
+
+impl<T, A, B, C> ManagedTask for DataSetDownloaderDaemon<T, A, B, C>
+where
+    T: DataSet,
+    A: HexAssignment,
+    B: HexAssignment,
+    C: HexAssignment,
+{
+    fn start_task(
+        self: Box<Self>,
+        shutdown: triggered::Listener,
+    ) -> futures::prelude::future::LocalBoxFuture<'static, anyhow::Result<()>> {
+        let handle = tokio::spawn(async move {
+            #[rustfmt::skip]
+            tokio::select! {
+                biased;
+                _ = shutdown.clone() => Ok(()),
+                result = self.run() => result,
+            }
+        });
+        Box::pin(
+            handle
+                .map_err(anyhow::Error::from)
+                .and_then(|result| async move { result.map_err(anyhow::Error::from) }),
+        )
+    }
+}
+
+impl<T, A, B, C> DataSetDownloaderDaemon<T, A, B, C>
```

Comment:
Can we get some tests around this?

Reply:
Unfortunately that's rather difficult right now. We don't really have any way to mock file stores, and a lot of this relies on the file store. 

I will do my best to test the things that are testable.